### ISSUE TITLE
Option to set grads of unused inputs to zeros instead of None

### ISSUE THREAD

**Issue by mfkasim1** [pytorch/pytorch#44189]
## 🚀 Feature
An option to set gradients of unused inputs to zeros instead of `None` in `torch.autograd.grad`.
Probably something like: `torch.autograd.grad(outputs, inputs, ..., zero_grad_unused=False)` where `zero_grad_unused` will be ignored if `allow_unused=False`.
If `allow_unused=True` and `zero_grad_unused=True`, then the returned gradients for unused inputs are zeros with the same shapes as the inputs.

## Motivation

I am working with multiple differentiations and I will find this feature very useful.
Consider a case below to calculate the 3rd gradient of a given function:

    import torch

    def get_3rd_derx(a, x, func):
        y = func(a, x)
        dydx = torch.autograd.grad(y, x, create_graph=True)
        d2ydx2 = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True)
        d3ydx3 = torch.autograd.grad(d2ydx2, x, allow_unused=True)
        return d3ydx3

    def func1(a, x):
        return a * x

    def func2(a, x):
        return torch.cos(a * x)

    dtype = torch.float64
    a = torch.tensor(0.3, dtype=dtype).requires_grad_()
    x = torch.tensor(1.0, dtype=dtype).requires_grad_()
    d3func1 = get_3rd_derx(a, x, func1) # I would expect a zero, but this raises an error
    d3func2 = get_3rd_derx(a, x, func2)
    print(d3func1)
    print(d3func2)

The first derivative of `func1` w.r.t. `x` does not have dependency to `x` anymore, so the 2nd derivative of `func2` will be `None`, thus raises an error when calculating the 3rd derivative.
In contrast to `func2`, the n-th derivatives of `x` will always have dependency to `x`, so it runs perfectly well.

## Pitch

With the current status, designing a functional (i.e. a function that receives another function as an input) to calculate multiple derivatives like above is annoying, because there is no way automatically check if the input parameters are already detached in first derivative (e.g. `func1`).

If there is an option to set zero grad to unused inputs, the gradient of the detached inputs will return the **mathematically correct** result (which is 0).

## Alternatives

Currently a workaround is possible without this feature, but it requires setting the `None` gradients manually to `0`, but it makes the code uglier.

## Additional context

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by dw-wang** [comment]
In tensorflow, they have the option to choose between:
UnconnectedGradients.ZERO
or
UnconnectedGradients.NONE
Some similar would be great!

**Comment by mfkasim1** [comment]
Any opinion on this @albanD?

**Comment by albanD** [comment]
In the context of custom Function, we use `materialize_grads` to do similar things.
I think that is ok to add as long as we implement this in the python code.

**Comment by zqigolden** [comment]
Hi, I'm interested in working on this task. Could you please assign it to me? Thanks!
